sdk/helper/ocsp/client.go

```diff
@@ -0,0 +1,1085 @@
+				if ret == nil {
+					// We may want to use this as the overall result
+					ret = ocspStatuses[i]
+					ocspRes = ocspResponses[i]
+				}
+			}
+		}
+	}
+
+	// If no server reported the cert revoked, but we did have an error, report it
+	if (ret == nil || ret.code == ocspStatusUnknown) && firstError != nil {
+		return nil, firstError
+	}
+	// otherwise ret should contain a response for the overall request
+
+	if !isValidOCSPStatus(ret.code) {
+		return ret, nil
+	}
+	v := ocspCachedResponse{
+		time:       float64(time.Now().UTC().Unix()),
+		producedAt: float64(ocspRes.ProducedAt.UTC().Unix()),
+		thisUpdate: float64(ocspRes.ThisUpdate.UTC().Unix()),
+		nextUpdate: float64(ocspRes.NextUpdate.UTC().Unix()),
+	}
+
+	c.ocspResponseCacheLock.Lock()
+	c.ocspResponseCache.Add(*encodedCertID, &v)
+	c.cacheUpdated = true
+	c.ocspResponseCacheLock.Unlock()
+	return ret, nil
+}
+
+func isValidOCSPStatus(status ocspStatusCode) bool {
+	return status == ocspStatusGood || status == ocspStatusRevoked || status == ocspStatusUnknown
+}
+
+type VerifyConfig struct {
+	OcspEnabled         bool
+	ExtraCas            []*x509.Certificate
+	OcspServersOverride []string
+	OcspFailureMode     FailOpenMode
+	QueryAllServers     bool
+}
+
+// VerifyLeafCertificate verifies just the subject against it's direct issuer
+func (c *Client) VerifyLeafCertificate(ctx context.Context, subject, issuer *x509.Certificate, conf *VerifyConfig) error {
+	results, err := c.GetRevocationStatus(ctx, subject, issuer, conf)
+	if err != nil {
+		return err
+	}
+	if results.code == ocspStatusGood {
+		return nil
+	} else {
+		serial := issuer.SerialNumber
+		serialHex := strings.TrimSpace(certutil.GetHexFormatted(serial.Bytes(), ":"))
+		if results.code == ocspStatusRevoked {
+			return fmt.Errorf("certificate with serial number %s has been revoked", serialHex)
+		} else if conf.OcspFailureMode == FailOpenFalse {
+			return fmt.Errorf("unknown OCSP status for cert with serial number %s", strings.TrimSpace(certutil.GetHexFormatted(serial.Bytes(), ":")))
+		} else {
+			c.Logger().Warn("could not validate OCSP status for cert, but continuing in fail open mode", "serial", serialHex)
+		}
+	}
+	return nil
+}
+
+// VerifyPeerCertificate verifies all of certificate revocation status
+func (c *Client) VerifyPeerCertificate(ctx context.Context, verifiedChains [][]*x509.Certificate, conf *VerifyConfig) error {
+	for i := 0; i < len(verifiedChains); i++ {
+		// Certificate signed by Root CA. This should be one before the last in the Certificate Chain
+		numberOfNoneRootCerts := len(verifiedChains[i]) - 1
+		if !verifiedChains[i][numberOfNoneRootCerts].IsCA || string(verifiedChains[i][numberOfNoneRootCerts].RawIssuer) != string(verifiedChains[i][numberOfNoneRootCerts].RawSubject) {
+			// Check if the last Non Root Cert is also a CA or is self signed.
+			// if the last certificate is not, add it to the list
+			rca := c.caRoot[string(verifiedChains[i][numberOfNoneRootCerts].RawIssuer)]
```

Comment:
Root subjects are not necessarily unique, but we can fix this later I suppose :-)